tools/cli/lib/api.c

```diff
@@ -1043,3 +1043,237 @@ TDNFCliRefresh(
     return TDNFRefresh(pContext->hTdnf);
 }
 
+static
+uint32_t
+TDNFCliHistoryAlter(
+    PTDNF_CLI_CONTEXT pContext,
+    PTDNF_CMD_ARGS pCmdArgs,
+    PTDNF_HISTORY_ARGS pHistoryArgs
+)
+{
+    uint32_t dwError = 0;
+    PTDNF_SOLVED_PKG_INFO pSolvedPkgInfo = NULL;
+
+    if(!pContext || !pCmdArgs || !pHistoryArgs)
+    {
+        dwError = ERROR_TDNF_CLI_INVALID_ARGUMENT;
+        BAIL_ON_CLI_ERROR(dwError);
+    }
+
+    dwError = pContext->pFnHistoryResolve(pContext, pHistoryArgs, &pSolvedPkgInfo);
+    BAIL_ON_CLI_ERROR(dwError);
+
+    if (pHistoryArgs->nCommand == HISTORY_CMD_INIT)
+    {
+        /* There is nothing to do here. */
+    }
+    else if (!(pSolvedPkgInfo->ppszPkgsNotResolved && pSolvedPkgInfo->ppszPkgsNotResolved[0]))
+    {
+        dwError = TDNFCliAskAndAlter(pContext, pCmdArgs, 0, pSolvedPkgInfo);
+        BAIL_ON_CLI_ERROR(dwError);
+    }
+    else
+    {
+        char *pszName = NULL;
+        int i;
+        pr_crit("The following packages could not be resolved:\n\n");
+        for (i = 0, pszName = pSolvedPkgInfo->ppszPkgsNotResolved[0];
+             pszName;
+             i++, pszName = pSolvedPkgInfo->ppszPkgsNotResolved[i])
+        {
+            pr_crit("%s\n", pszName);
+        }
+        pr_crit("\n"
+                "The package(s) may have been moved out of the enabled repositories since the\n"
+                "last time they were installed. You may be able to resolve this by enabling\n"
+                "additional repositories.\n");
+        dwError = ERROR_TDNF_NO_MATCH;
+        BAIL_ON_CLI_ERROR(dwError);
+    }
+
+cleanup:
+    TDNFCliFreeSolvedPackageInfo(pSolvedPkgInfo);
+    return dwError;
+error:
+    goto cleanup;
+}
+
+static
+uint32_t
+TDNFCliHistoryList(
+    PTDNF_CLI_CONTEXT pContext,
+    PTDNF_CMD_ARGS pCmdArgs,
+    PTDNF_HISTORY_ARGS pHistoryArgs
+)
+{
+    uint32_t dwError = 0;
+    PTDNF_HISTORY_INFO pHistoryInfo = NULL;
+    PTDNF_HISTORY_INFO_ITEM pItems = NULL;
+    struct json_dump *jd = NULL;
+    struct json_dump *jd_item = NULL;
+    struct json_dump *jd_list_added = NULL;
+    struct json_dump *jd_list_removed = NULL;
+
+    if(!pContext || !pCmdArgs || !pHistoryArgs)
+    {
+        dwError = ERROR_TDNF_CLI_INVALID_ARGUMENT;
+        BAIL_ON_CLI_ERROR(dwError);
+    }
+
+    dwError = pContext->pFnHistoryList(pContext, pHistoryArgs, &pHistoryInfo);
+    BAIL_ON_CLI_ERROR(dwError);
+
+    pItems = pHistoryInfo->pItems;
+
+    if (!pCmdArgs->nJsonOutput)
```

Comment:
can we put both if and else separately with thier respective function . Although it requires  two calling but looks more readable

Reply:
Hm, I don't agree that it would make it more readable. It's one more step if you want to follow the code.